### ISSUE TITLE
test.unit.dart failing on Windows

### ISSUE THREAD

**Issue by mlaval** [angular/angular#1875]
Here is what I get when running `gulp test.unit.dart`:

```
[18:44:34] Starting 'build/clean.dart'...
[18:44:34] Starting 'build/clean.tools'...
[18:44:34] Finished 'build/clean.tools' after 1.59 ms
[18:44:34] Starting 'build.tools'...
[18:44:34] Starting '!build.tools'...
[18:44:34] Finished 'build/clean.dart' after 113 ms
[18:44:38] Finished '!build.tools' after 4.43 s
[18:44:38] Finished 'build.tools' after 4.43 s
[18:44:38] Starting 'build/tree.dart'...
[18:44:38] Starting '!build/tree.dart'...
Tree diff:                   funnel-dest_XKNp9u.tmp, duration:    65ms,   398 changes detected (files:   398, directories:  217)

Slowest Trees                                 | Total
----------------------------------------------+---------------------
Funnel                                        | 926ms
DestCopy                                      | 764ms
Funnel                                        | 701ms
TreeMerger                                    | 677ms
Funnel                                        | 660ms
TreeMerger                                    | 409ms
Funnel                                        | 315ms

Slowest Trees (cumulative)                    | Total (avg)
----------------------------------------------+---------------------
Funnel (12)                                   | 3388ms (282 ms)
TreeMerger (5)                                | 1784ms (356 ms)
DestCopy (1)                                  | 764ms

[18:44:44] Finished '!build/tree.dart' after 6.34 s
[18:44:44] Finished 'build/tree.dart' after 6.34 s
[18:44:44] Starting 'test.unit.dart'...
[18:44:44] Starting '!test.unit.dart/karma-server'...
[18:44:45] Finished '!test.unit.dart/karma-server' after 303 ms
[18:44:45] Starting '!test.unit.dart/karma-run'...
INFO [karma]: Karma v0.12.31 server started at http://localhost:9877/
INFO [launcher]: Starting browser Dartium
No captured browser, open http://localhost:9877/
[18:44:45] Finished '!test.unit.dart/karma-run' after 121 ms
INFO [Chrome 39.0.2171 (Windows 7)]: Connected on socket Vmy3Az9So-ZTegeeKgDC with id 83873509
WARN [web-server]: 404: /base/dist/dart/angular2/lib/angular2.dart
WARN [web-server]: 404: /base/dist/dart/angular2/lib/di.dart
WARN [web-server]: 404: /base/dist/dart/angular2/lib/src/test_lib/test_bed.dart
WARN [web-server]: 404: /base/dist/dart/angular2/lib/test_lib.dart
```

The transpilation from TS to Dart is failing, issue opened in ts2dart repo: https://github.com/angular/ts2dart/issues/157

**Comment by mlaval** [comment]
Still failing but with another error:
```
[23:18:29] '!build/tree.dart' errored after 5.13 s
[23:18:29] Error: spawn ENAMETOOLONG
    at exports._errnoException (util.js:746:11)
    at ChildProcess.spawn (child_process.js:1162:11)
    at exports.spawn (child_process.js:995:9)
    at Object.exports.execFile (child_process.js:689:15)
    at exports.exec (child_process.js:649:18)
    at c:\Dev\angular\angular_mlaval\broccoli-dartfmt.ts:49:7
    at DartFormatter.rebuild (c:\Dev\angular\angular_mlaval\broccoli-dartfmt.ts:47:12)
    at DiffingPluginWrapper.rebuild (c:\Dev\angular\angular_mlaval\diffing-broccoli-plugin.ts:87:47)
    at c:\Dev\angular\angular_mlaval\node_modules\broccoli\lib\api_compat.js:42:21
    at lib$rsvp$$internal$$tryCatch (c:\Dev\angular\angular_mlaval\node_modules\broccoli\node_modules\rsvp\dist\rsvp.js:489:16)
    at lib$rsvp$$internal$$invokeCallback (c:\Dev\angular\angular_mlaval\node_modules\broccoli\node_modules\rsvp\dist\rsvp.js:501:17)
    at c:\Dev\angular\angular_mlaval\node_modules\broccoli\node_modules\rsvp\dist\rsvp.js:1095:13
    at lib$rsvp$asap$$flush (c:\Dev\angular\angular_mlaval\node_modules\broccoli\node_modules\rsvp\dist\rsvp.js:1290:9)
```

It seems to come from the dartfmt command which in that case is 35614 characters long, while 8191 is the limit in Windows command line.
The command starts with:
`dartfmt.bat -w c:\Dev\angular\angular_mlaval\tmp\diffing_plugin_wrapper-cache_path-yHAKKF9H.tmp\angular2\angular2.api.dart c:\Dev\angular\angular_mlaval\tmp\diffing_plugin_wrapper-cache_path-yHAKKF9H.tmp\angular2\angular2_sfx.dart c:\Dev\an...`

**Comment by caitp** [comment]
may need to quote each filename, could you try that and see if it's an easy fix?

**Comment by mlaval** [comment]
It doesn't help

**Comment by caitp** [comment]
try that PR's patch and see if that does it for you

**Comment by mlaval** [comment]
It works, great!

Now the task fails later:
```
[00:59:48] Starting '!build/change_detect.dart'...
[00:59:48] '!build/change_detect.dart' errored after 6.65 ms
[00:59:48] Error: Failed while generating change detector classes. Please run manually: dart.bat
    at ChildProcess.<anonymous> (c:\Dev\angular\angular_mlaval\gulpfile.js:864:10)
    at ChildProcess.emit (events.js:107:17)
    at Process.ChildProcess._handle.onexit (child_process.js:1072:12)
    at child_process.js:1144:20
    at process._tickCallback (node.js:355:11)
[00:59:48] 'test.unit.dart' errored after 29 s
[00:59:48] Error: [object Object]
    at formatError (c:\Users\Marc\AppData\Roaming\npm\node_modules\gulp\bin\gulp.js:169:10)
    at Gulp.<anonymous> (c:\Users\Marc\AppData\Roaming\npm\node_modules\gulp\bin\gulp.js:195:15)
    at Gulp.emit (events.js:129:20)
    at Gulp.Orchestrator._emitTaskDone (c:\Dev\angular\angular_mlaval\node_modules\gulp\node_modules\orchestrator\index.js:264:8)
    at c:\Dev\angular\angular_mlaval\node_modules\gulp\node_modules\orchestrator\index.js:275:23
    at finish (c:\Dev\angular\angular_mlaval\node_modules\gulp\node_modules\orchestrator\lib\runTask.js:21:8)
    at cb (c:\Dev\angular\angular_mlaval\node_modules\gulp\node_modules\orchestrator\lib\runTask.js:29:3)
    at c:\Dev\angular\angular_mlaval\gulpfile.js:444:9
    at finish (c:\Dev\angular\angular_mlaval\node_modules\run-sequence\index.js:53:5)
    at Gulp.onError (c:\Dev\angular\angular_mlaval\node_modules\run-sequence\index.js:60:4)
    at Gulp.emit (events.js:129:20)
    at Gulp.Orchestrator._emitTaskDone (c:\Dev\angular\angular_mlaval\node_modules\gulp\node_modules\orchestrator\index.
js:264:8)
    at c:\Dev\angular\angular_mlaval\node_modules\gulp\node_modules\orchestrator\index.js:275:23
    at finish (c:\Dev\angular\angular_mlaval\node_modules\gulp\node_modules\orchestrator\lib\runTask.js:21:8)
    at cb (c:\Dev\angular\angular_mlaval\node_modules\gulp\node_modules\orchestrator\lib\runTask.js:29:3)
    at ChildProcess.<anonymous> (c:\Dev\angular\angular_mlaval\gulpfile.js:864:5)
    at ChildProcess.emit (events.js:107:17)
    at Process.ChildProcess._handle.onexit (child_process.js:1072:12)
    at child_process.js:1144:20
    at process._tickCallback (node.js:355:11)
Error: ENOENT, open 'c:\Dev\angular\angular_mlaval\dist\dart\angular2\test\change_detection\generated\simple_watch_classes.dart'
    at Error (native)
```

To be continued

**Comment by mlaval** [comment]
Thanks for the help @caitp 
With this improvement, there are still some failures happening. Working on it

**Comment by piloopin** [comment]
I had some progress on building dart version on windows. Here are some issues:

1. The later issue is because of https://github.com/angular/angular/blob/c7572ac1f9bbda59ba301dd330e6f64faf91b9e1/tools/build/dartdetect.js#L13. It should be dart.exe.
2. https://github.com/angular/ts2dart/issues/191
3. https://github.com/angular/angular/blob/c7572ac1f9bbda59ba301dd330e6f64faf91b9e1/tools/build/dartanalyzer.js#L27 should be `fileName = 'package:' + packageName + '/' + path.relative('lib', fileName).replace(/\\/g, "/");`
4. https://github.com/angular/angular/blob/c7572ac1f9bbda59ba301dd330e6f64faf91b9e1/tools/build/pubbuild.js#L51 should be `fs.renameSync(subFolder, subFolder.replace('/web/', '/'));` I don't know why the slashes are normalized here, but they are.

But I stuck at pubbuild step. I seems that angular dart transformer is not dealing with windows backslashes very well.

**Comment by mlaval** [comment]
Nice to see that I'm not alone on this quest :) 
I had identified the first 2 issues too, for the same reasons.

A follow up of https://github.com/angular/ts2dart/issues/191 would be to normalize slashes in https://github.com/angular/angular/blob/master/tools/broccoli/broccoli-ts2dart.ts
Depending how `ts2dart` is fixed, it may concern lines 18, 34 and 47.

**Comment by kegluneq** [comment]
Thanks for the effort on this and sorry for my part in the trouble (I've been working on the angular 2 dart transformer). Is there a stack trace I can look at re: the pub build step?

**Comment by mlaval** [comment]
Fixed by #2676 and https://github.com/angular/ts2dart/pull/224 which was released in ts2dart 0.6.6
Working fine now!
